instrgen/docs/design.md

```diff
@@ -0,0 +1,200 @@
+
+```go
+// PackageRewriter interface does actual input package
+// rewriting according to specific criteria.
+type PackageRewriter interface {
+        // ID Dumps rewriter id.
+        Id() string
+        // Inject tells whether package should be rewritten.
+        Inject(pkg string, filepath string) bool
+        // ReplaceSource decides whether input sources should be replaced
+        // or all rewriting work should be done in temporary location.
+        ReplaceSource(pkg string, filePath string) bool
+        // Rewrite does actual package rewriting.
+        Rewrite(pkg string, file *ast.File, fset *token.FileSet, trace *os.File)
+        // WriteExtraFiles generate additional files that will be linked
+        // together to input package.
+        // Additional files have to be returned as array of file names.
+        WriteExtraFiles(pkg string, destPath string) []string
+}
+```
+
+
+#### Example rewriter
+
+```go
+// FunDumper rewrites all functions according to FilePattern.
+type FunDumper struct {
+}
+
+// Id.
+func (FunDumper) Id() string {
+        return "FunDumper"
+}
+
+// Inject.
+func (fd FunDumper) Inject(pkg string, filepath string) bool {
+	return true
+}
+
+// ReplaceSource.
+func (fd FunDumper) ReplaceSource(pkg string, filePath string) bool {
+        return true
+}
+
+func makeStmts(name string) []ast.Stmt
+	s := &ast.ExprStmt {
+		X: &ast.CallExpr {
+			Fun: &ast.SelectorExpr {
+				X: &ast.Ident {
+					Name: "__myfmt",
+				},
+				Sel: &ast.Ident {
+					Name: "Println",
+				},
+			},
+			Lparen: 42,
+			Args: []ast.Expr {
+				&ast.BasicLit {
+					ValuePos: 43,
+					Kind: STRING,
+					Value: "\" + name +\"",
+				},
+			},
+			Ellipsis: 0,
+		},
+	}
+        return []ast.Stmt{s}
+
+// Rewrite.
+func (fd FunDumper) Rewrite(pkg string, file *ast.File, fset *token.FileSet, trace *os.File) {
+        visited := make(map[string]bool, 0)
+        ast.Inspect(file, func(n ast.Node) bool {
+                if funDeclNode, ok := n.(*ast.FuncDecl); ok {
+                        // check if functions has been already instrumented
+                        if _, ok := visited[fset.Position(file.Pos()).String()+":"+funDeclNode.Name.Name]; !ok {
+                                funDeclNode.Body.List = append(makeStmts(funDeclNode.Name.Name), funDeclNode.Body.List...)
+                                astutil.AddNamedImport(fset, file, "__myfmt", "fmt")
+                                visited[fset.Position(file.Pos()).String()+":"+funDeclNode.Name.Name] = true
+                        }
+                }
+                return true
+        })
+}
+
+// WriteExtraFiles.
+func (FunDumper) WriteExtraFiles(pkg string, destPath string) []string {
+        return nil
+}
+```
+
+### UI
+
+To lower usage barrier to user `instrgen` offers simple UI where instrumentations of specific functions and libraries
+can be selected.
+
+![image info](./instrgen.png)
+
+### Scope
```

Comment:
Is context plumbing between instrumentation in-scope?

Is SDK configuration in-scope?  What will the configuration look like?

Reply:
So, for now I was thinking about enabling/disabling instrumentation of specific libraries or functions. Currently plumbing and SDK configuration was out of scope, however that might evolve.

Reply:
What is the user journey you are proposing, then for the compile-time instrumentation?  I was imagining something like:

### Compile-time User journey

1. User has application without any instrumentation
2. User uses this tool as part of their CI with some configuration for what is instrumented
3. User configures telemetry export using environment variables or the (coming soon) config file.

But that only works if it can do context plumbing and SDK configuration